### PR TITLE
Lets user specify optional packages at build time

### PR DESCRIPTION
**Pull request by zigahertz**
This in response to https://github.com/bgentry/ember-apollo-client/issues/108.

This *almost* works - one problem is that I also need to explicitly install whatever extra packages I've specifed, for example, if `config/environment.js` looks like:

```
apollo: {
  ...
  include: ['apollo-link-batch-http', 'apollo-link-persisted-queries'],
}

```

My app complains about missing modules unless I do
```
$ npm i --save-dev apollo-link-batch-http apollo-link-persisted-queries
```

So far I haven't figured out how to include these in an additional blueprint... any ideas?

### CODE REVIEW

**Comment by ivanvanderbyl** [comment]
Damn that spread operator. @bgentry how do you feel about only supporting newer node than 4?

**Comment by viniciussbs** [comment]
What do you think about following the [`ember-cli-fastboot` approach](https://github.com/ember-fastboot/ember-cli-fastboot#whitelisting-packages)?

```javascript
// package.json
{
  "name": "my-sweet-app",
  "version": "0.4.2",
  "devDependencies": {
    "apollo-link-batch-http",
    "apollo-link-persisted-queries"
    // ...
  },
  "dependencies": {
    // ...
  },
  "apolloClientDependencies": [
    "apollo-link-batch-http",
    "apollo-link-persisted-queries"
  ]
}
```

[The implementation](https://github.com/ember-fastboot/ember-cli-fastboot/blob/88c646394bd9575cbac3fbb923f4c2ba0fa42561/lib/broccoli/fastboot-config.js), I guess.

File: README.md [hunk 1]
```diff
@@ -53,13 +53,15 @@ The application built in the tutorial is also available on the [How To GraphQL r
...
-In your app's `config/environment.js`, configure the URL for the GraphQL API:
+In your app's `config/environment.js`, configure the URL for the GraphQL API. Here, you can also include additional Apollo packages that haven't been included by default. Valid package names can be found [here](https://github.com/apollographql/apollo-link/tree/master/packages)
+
 
 ```js
 let ENV = {
   ...
   apollo: {
     apiURL: 'https://test.example/graphql',
+    include: ['apollo-link-batch-http', 'apollo-link-persisted-queries'],
```
Comment (villander) [inline README.md]: Whitelist packages is a approach better and more funded in `ember/addons` ecosystem. 

https://github.com/ember-fastboot/ember-cli-fastboot#whitelisting-packages

File: index.js [hunk 2]
```diff
@@ -1,34 +1,40 @@
...
 module.exports = {
   name: 'ember-apollo-client',
 
-  included() {
+  included(app) {
     this._super.included.apply(this, arguments);
-
+    
+    let config = app.project.config(app.env) || {};
+    this.addonConfig = config['apollo'] || {};
```
Comment (villander) [inline index.js]: This is more sugar

```js
this.addonConfig = this.app.project.config(app.env)['apollo'] || {};
```

File: index.js [hunk 3]
```diff
@@ -1,34 +1,40 @@
...
+    
     this.import('vendor/-apollo-client-bundle.js');
     this.import('vendor/-apollo-client-shims.js');
   },
-
+  
   treeForVendor() {
     const WebpackDependencyPlugin = require('./lib/webpack-dependency-plugin');
-
+    const userSpecifiedPackages = this.addonConfig.include;
```
Comment (villander) [inline index.js]: `const apolloSpecifiedPackages ` is a better name.

File: index.js [hunk 4]
```diff
@@ -1,34 +1,40 @@
...
-        'apollo-client',
-        'apollo-link',
-        'apollo-link-context',
-        'apollo-link-http',
-        'graphql',
-        'graphql-tools',
-        'graphql-tag'
-      ]
-    });
+      expose: [...defaultPackages, ...userSpecifiedPackages],
```
Comment (bgentry) [inline index.js]: What do you think about having a default list of packages, but then if the user specifies any packages we *only* use the ones they've specified? The goal would be to enable users to remove as well as add packages.

File: index.js [hunk 5]
```diff
@@ -1,34 +1,40 @@
...
-
+    
+    let config = app.project.config(app.env) || {};
+    this.addonConfig = config['apollo'] || {};
+    
     this.import('vendor/-apollo-client-bundle.js');
     this.import('vendor/-apollo-client-shims.js');
   },
-
+
```
Comment (bgentry) [inline index.js]: no trailing whitespace please :)

File: index.js [hunk 6]
```diff
@@ -1,34 +1,40 @@
...
-        'graphql',
-        'graphql-tools',
-        'graphql-tag'
-      ]
-    });
+      expose: [...defaultPackages, ...userSpecifiedPackages],
+    })
   },
-
+
```
Comment (bgentry) [inline index.js]: no trailing whitespace please :)

File: index.js (hunk 2 above)
Comment (bgentry) [reply index.js]: Is there a case where `this.app.project.config(app.env)` could be `undefined`? I don't feel strongly about this though, the existing way is a bit easier to skim.

File: README.md [hunk 7]
```diff
@@ -72,6 +74,7 @@ Additional configuration of the ApolloClient can be done by extending the Apollo
 service and overriding the `clientOptions` property. See the
 [Apollo Service API][apollo-service-api] for more info.
 
+
```
Comment (bgentry) [inline README.md]: Can you get rid of this extra newline you added?

File: README.md (hunk 1 above)
Comment (dfreeman) [reply README.md]: The addon ecosystem isn't _perfect_ about following these conventions, but the rough guidance I've seen is that config that affects runtime stuff goes in `config/environment.js`, while build-time stuff goes in `ember-cli-build.js` (or `index.js` for addons). Given that, I'd vote in favor of the latter.

File: index.js (hunk 2 above)
Comment (villander) [reply index.js]: No! This is the better way, `ember-cli-mirage` use like this.

File: README.md (hunk 1 above)
Comment (villander) [reply README.md]: I agree `build-time` seems a better way, it makes sense to me.

**Comment by bgentry** [comment]
Hi @zigahertz! I'd really love to get this in before a proper 1.0 release, but I've already been waiting quite awhile to do that. Any chance you'll have time to make another pass at this soon?

**Comment by zigahertz** [comment]
@bgentry - yes I'll make updates and push today!

**Comment by zigahertz** [comment]
@bgentry - 

1. I added the ability to explicitly exclude packages, I feel that's a bit more clearer and less verbose than the alternative we were discussing. 

2. README has been updated with instructions

**Comment by bgentry** [comment]
@zigahertz hey, really sorry for not making this clearer earlier. Based on [the prior discussion](https://github.com/bgentry/ember-apollo-client/pull/109/files#r175642471) around where the build config should go, it seems clear to me that `ember-cli-build.js` is the more logical place.

Based on the implementations in [ember-fontawesome](https://github.com/FortAwesome/ember-fontawesome/blob/master/index.js#L79-L84) and [ember-bootstrap](https://github.com/kaliber5/ember-bootstrap/blob/f182a291c5468dea386e8a1ff69e7ebc7e1209f7/index.js#L77) it seems like it might be pretty easy to make this change. I think it's essentially just a matter of accessing `this.app.options.apollo` to get access to the build config hash rather than using the `env`.

Is this a change you'd be able to make? Otherwise I think I will probably end up doing it myself before getting this merged & released.

**Comment by bgentry** [comment]
To be clear, I think this looks really great otherwise! Thanks a bunch for all the small readme fixes you took care of too :pray:

**Comment by zigahertz** [comment]
@bgentry good to hear! and, yes, definitely, happy to make those changes later this morning -- should have those pushed up by around 10A PST.

**Comment by zigahertz** [comment]
@bgentry ok, there you go - readme updated as well. What do you think?

**Comment by bgentry** [comment]
Awesome stuff @zigahertz, thank you! I've gone ahead and done a manual merge on this to squash it down but keep the Node v6 commit separate.

Feel free to open a new PR to add yourself to the readme contributors list 😄

🎉🚀✌️

**Comment by zigahertz** [comment]
@bgentry awesome!

**Pull request closed without merging**